### ISSUE TITLE
Still got error when using BatchNormalization layer

### ISSUE THREAD

**Issue by dongb5** [fchollet/keras#6010]
When I added a BatchNormalization layer in my architecture definition, I got a error as follows:

` InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'input_15' with dtype float
	 [[Node: input_15 = Placeholder[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
	 [[Node: batch_normalization_5/moments/sufficient_statistics/Shape/_551 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_1060_batch_normalization_5/moments/sufficient_statistics/Shape", tensor_type=DT_INT32, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]`

I have eliminated the error by set mode==2. However, the 'mode' param of BatchNormalization layer is cancelled in Keras2. Now, the error comes back! Any suggestion?

**Comment by Namnamseo** [comment]
Can you upload your architecture?

**Comment by dongb5** [comment]
@Namnamseo 

http://stackoverflow.com/questions/42422646/keras-train-partial-model-issue-about-gan-model

The problem shown in the above url is exactly what I have met. I had this problem when training Wasserstein GAN, and I am pretty sure that the BatchNormalization layer caused the problem. Any help?

**Comment by Namnamseo** [comment]
I suppose the batch size is related here.

In Keras, when building the model, the size of the batch axis is unknown. And as Batch Normalization layer in Keras calculates the statistics it needs by dividing sums by the number of nodes, it needs to know the (exact, multiplied by the batch size) number of nodes. The number of nodes is calculated just like other operations; `tf.shape()` function is called on the input, and all values of it are multiplied.

Thus, there is a 'flow' from the input *dimensions* to the output *values*. And for the input *dimensions*, specifically the batch size, it *flows* from G to D. In other words, the batch size of G becomes the batch size of D, as we combined G's output to D's input when building GAN model.

This is true even without the Batch Normalization layer; the two would have the same batch size with GAN model. But Batch Normalization layer explicitly calculates the number of nodes, by accessing the dimensions, affecting the calculation graph.

**Why functional APIs solve this?** I suspect that the Sequential model (especially when compiled?) arranges the inputs and outputs, colliding somehow with what Batch Normalization layer did.

**Why `mode==2` solve this?** I'm not sure because I hasn't looked deep into the issue (I haven't even run the source you mentioned yet). As the difference between the modes is the moving average and `uses_learning_phase`, I guess one of the two affects the things I mentioned above.

This was my guess about what's happening. For the solution, I suggest using functional API; stacking some specifically designed(sequential for here) models to form a new model seems a little unstable.

**Comment by dongb5** [comment]
Thanks for the suggestions. I have already used functional API to build the models, however, the problem remains. I'm tring to follow the answer of the question in stackoverflow which is to define every layer in discriminator. Hope it will work. Thanks again anyway!

**Comment by lijin-THU** [comment]
I've encountered the same problem and I solved it by set every layer of my discriminator trainable as False.
```python
D.trainable = False
for layer in D.layers:
     layer.trainable = False
```
Maybe you can try this.
